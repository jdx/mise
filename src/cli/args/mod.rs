pub use cd_arg::CdArg;
pub use env_var_arg::EnvVarArg;
pub use forge_arg::ForgeArg;
pub use log_level_arg::{DebugArg, LogLevelArg, TraceArg};
pub use quiet_arg::QuietArg;
pub use tool_arg::ToolArg;
pub use tool_arg::ToolVersionType;
pub use verbose_arg::VerboseArg;
pub use yes_arg::YesArg;

mod cd_arg;
mod env_var_arg;
mod forge_arg;
mod log_level_arg;
mod quiet_arg;
mod tool_arg;
mod verbose_arg;
mod yes_arg;
