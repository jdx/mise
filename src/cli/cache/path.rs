use eyre::Result;

use crate::env;

/// Show the cache directory path
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, visible_alias = "dir")]
pub struct CachePath {}

impl CachePath {
    pub fn run(self) -> Result<()> {
        miseprintln!("{}", env::MISE_CACHE_DIR.display());
        Ok(())
    }
}
