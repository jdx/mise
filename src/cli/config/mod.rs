use clap::Subcommand;
use eyre::Result;

mod generate;
mod ls;

/// [experimental] Manage config files
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "cfg")]
pub struct Config {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// Do not print table header
    #[clap(long, alias = "no-headers", verbatim_doc_comment)]
    no_header: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Ls(ls::ConfigLs),
    Generate(generate::ConfigGenerate),
}

impl Commands {
    pub fn run(self) -> Result<()> {
        match self {
            Self::Ls(cmd) => cmd.run(),
            Self::Generate(cmd) => cmd.run(),
        }
    }
}

impl Config {
    pub fn run(self) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::ConfigLs {
            no_header: self.no_header,
        }));

        cmd.run()
    }
}
