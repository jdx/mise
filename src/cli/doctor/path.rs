use crate::config::Config;
use crate::Result;
use std::env;

/// Print the current PATH entries mise is providing
#[derive(Debug, clap::Args)]
#[clap(alias="paths", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Path {
    /// Print all entries including those not provided by mise
    #[clap(long, short, verbatim_doc_comment)]
    full: bool,
}

impl Path {
    pub fn run(self) -> Result<()> {
        let config = Config::get();
        let ts = config.get_toolset()?;
        let paths = if self.full {
            let env = ts.env_with_path(&config)?;
            let path = env.get("PATH").cloned().unwrap_or_default();
            env::split_paths(&path).collect()
        } else {
            ts.list_final_paths()?
        };
        for path in paths {
            println!("{}", path.display());
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    Get the current PATH entries mise is providing
    $ mise path
    /home/user/.local/share/mise/installs/node/24.0.0/bin
    /home/user/.local/share/mise/installs/rust/1.90.0/bin
    /home/user/.local/share/mise/installs/python/3.10.0/bin
"#
);
