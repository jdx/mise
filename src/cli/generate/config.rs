use crate::Result;
use crate::cli::config::generate;

/// [experimental] Generate a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = generate::AFTER_LONG_HELP)]
pub struct Config {
    #[clap(flatten)]
    generate: generate::ConfigGenerate,
}

impl Config {
    pub async fn run(self) -> Result<()> {
        self.generate.run()
    }
}
