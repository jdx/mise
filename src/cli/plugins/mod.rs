use clap::Subcommand;
use eyre::Result;

use crate::config::Config;

mod install;
mod link;
mod ls;
mod ls_remote;
mod uninstall;
mod update;

#[derive(Debug, clap::Args)]
#[clap(about = "Manage plugins", visible_alias = "p", aliases = ["plugin", "plugin-list"])]
pub struct Plugins {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// list all available remote plugins
    ///
    /// same as `rtx plugins ls-remote`
    #[clap(short, long, hide = true)]
    pub all: bool,

    /// The built-in plugins only
    /// Normally these are not shown
    #[clap(short, long, verbatim_doc_comment, conflicts_with = "all")]
    pub core: bool,

    /// List installed plugins
    ///
    /// This is the default behavior but can be used with --core
    /// to show core and user plugins
    #[clap(long, verbatim_doc_comment, conflicts_with = "all")]
    pub user: bool,

    /// show the git url for each plugin
    ///
    /// e.g.: https://github.com/asdf-vm/asdf-node.git
    #[clap(short, long)]
    pub urls: bool,

    /// show the git refs for each plugin
    ///
    /// e.g.: main 1234abc
    #[clap(long)]
    pub refs: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Install(install::PluginsInstall),
    Link(link::PluginsLink),
    Ls(ls::PluginsLs),
    LsRemote(ls_remote::PluginsLsRemote),
    Uninstall(uninstall::PluginsUninstall),
    Update(update::Update),
}

impl Commands {
    pub fn run(self, config: &Config) -> Result<()> {
        match self {
            Self::Install(cmd) => cmd.run(config),
            Self::Link(cmd) => cmd.run(),
            Self::Ls(cmd) => cmd.run(config),
            Self::LsRemote(cmd) => cmd.run(config),
            Self::Uninstall(cmd) => cmd.run(config),
            Self::Update(cmd) => cmd.run(config),
        }
    }
}

impl Plugins {
    pub fn run(self, config: &Config) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::PluginsLs {
            all: self.all,
            core: self.core,
            refs: self.refs,
            urls: self.urls,
            user: self.user,
        }));

        cmd.run(config)
    }
}
