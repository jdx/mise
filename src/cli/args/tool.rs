use std::ffi::{OsStr, OsString};
use std::fmt::Display;

use clap::{Arg, Command, Error};
use color_eyre::eyre::Result;
use regex::Regex;

use crate::plugins::PluginName;
use crate::toolset::ToolVersionRequest;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ToolArg {
    pub plugin: PluginName,
    pub tvr: Option<ToolVersionRequest>,
}

impl ToolArg {
    pub fn parse(input: &str) -> Self {
        match input.split_once('@') {
            Some((plugin, version)) => Self {
                plugin: plugin.to_string(),
                tvr: Some(ToolVersionRequest::new(plugin.to_string(), version)),
            },
            None => Self {
                plugin: input.into(),
                tvr: None,
            },
        }
    }

    /// this handles the case where the user typed in:
    /// rtx local node 20.0.0
    /// instead of
    /// rtx local node@20.0.0
    ///
    /// We can detect this, and we know what they meant, so make it work the way
    /// they expected.
    pub fn double_tool_condition(tools: &[ToolArg]) -> Vec<ToolArg> {
        let mut tools = tools.to_vec();
        if tools.len() == 2 {
            let re: &Regex = regex!(r"^\d+(\.\d+)?(\.\d+)?$");
            let a = tools[0].clone();
            let b = tools[1].clone();
            if matches!(a.tvr, None) && matches!(b.tvr, None) && re.is_match(&b.plugin) {
                tools[1].tvr = Some(ToolVersionRequest::new(a.plugin.clone(), &b.plugin));
                tools[1].plugin = a.plugin;
                tools.remove(0);
            }
        }
        tools
    }

    pub fn with_version(self, version: &str) -> Self {
        Self {
            tvr: Some(ToolVersionRequest::new(self.plugin.clone(), version)),
            ..self
        }
    }
}

impl Display for ToolArg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.tvr {
            Some(tvr) => write!(f, "{}", tvr),
            _ => write!(f, "{}", self.plugin),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ToolArgParser;

impl clap::builder::TypedValueParser for ToolArgParser {
    type Value = ToolArg;

    fn parse_ref(
        &self,
        cmd: &Command,
        arg: Option<&Arg>,
        value: &OsStr,
    ) -> Result<Self::Value, Error> {
        self.parse(cmd, arg, value.to_os_string())
    }

    fn parse(
        &self,
        _cmd: &Command,
        _arg: Option<&Arg>,
        value: OsString,
    ) -> Result<Self::Value, Error> {
        Ok(ToolArg::parse(&value.to_string_lossy()))
    }
}
