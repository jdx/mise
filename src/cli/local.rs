use color_eyre::eyre::{eyre, ContextCompat, Result};

use crate::cli::args::runtime::{RuntimeArg, RuntimeArgParser};
use crate::cli::command::Command;
use crate::config::{config_file, Config};
use crate::output::Output;
use crate::plugins::PluginName;
use crate::{dirs, env, file};

/// Sets .tool-versions to include a specific runtime
///
/// use this to set the runtime version when within a directory
/// use `rtx global` to set a runtime version globally
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, visible_alias = "l", after_long_help = AFTER_LONG_HELP)]
pub struct Local {
    /// runtimes to add to .tool-versions
    ///
    /// e.g.: nodejs@20
    #[clap(value_parser = RuntimeArgParser)]
    runtime: Option<Vec<RuntimeArg>>,

    /// recurse up to find a .tool-versions file rather than using the current directory only
    /// by default this command will only set the runtime in the current directory ("$PWD/.tool-versions")
    #[clap(short, long)]
    parent: bool,

    /// save fuzzy match to .tool-versions
    /// e.g.: `rtx local --fuzzy nodejs@20` will save `nodejs 20` to .tool-versions
    /// by default it would save the exact version, e.g.: `nodejs 20.0.0`
    #[clap(long)]
    fuzzy: bool,

    /// remove the plugin(s) from .tool-versions
    #[clap(long, value_name = "PLUGIN")]
    remove: Option<Vec<PluginName>>,
}

impl Command for Local {
    fn run(self, config: Config, out: &mut Output) -> Result<()> {
        let cf_path = match self.parent {
            true => file::find_up(
                &dirs::CURRENT,
                &[env::RTX_DEFAULT_TOOL_VERSIONS_FILENAME.as_str()],
            )
            .with_context(|| {
                eyre!(
                    "no {} file found",
                    env::RTX_DEFAULT_TOOL_VERSIONS_FILENAME.as_str()
                )
            })?,
            false => dirs::CURRENT.join(env::RTX_DEFAULT_TOOL_VERSIONS_FILENAME.as_str()),
        };

        let mut cf = match cf_path.exists() {
            true => config_file::parse(&cf_path)?,
            false => config_file::init(&cf_path),
        };

        if let Some(plugins) = &self.remove {
            for plugin in plugins {
                cf.remove_plugin(plugin);
            }
        }
        if let Some(runtimes) = &self.runtime {
            cf.add_runtimes(&config, runtimes, self.fuzzy)?;
        }

        if self.runtime.is_some() || self.remove.is_some() {
            cf.save()?;
        }

        rtxprint!(out, "{}", cf.dump());

        Ok(())
    }
}

const AFTER_LONG_HELP: &str = r#"
Examples:
  # set the current version of nodejs to 20.x for the current directory
  # will use a precise version (e.g.: 20.0.0) in .tool-versions file
  $ rtx local nodejs@20

  # set nodejs to 20.x for the current project (recurses up to find .tool-versions)
  $ rtx local -p nodejs@20

  # set the current version of nodejs to 20.x for the current directory
  # will use a fuzzy version (e.g.: 20) in .tool-versions file
  $ rtx local --fuzzy nodejs@20

  # removes nodejs from .tool-versions
  $ rtx local --remove=nodejs
"#;

#[cfg(test)]
mod test {
    use std::fs;

    use insta::assert_snapshot;
    use pretty_assertions::assert_str_eq;

    use crate::cli::test::grep;

    use crate::{assert_cli, dirs};

    #[test]
    fn test_local() {
        let cf_path = dirs::CURRENT.join(".tool-versions");
        let orig = fs::read_to_string(&cf_path).unwrap();

        assert_cli!("plugin", "add", "nodejs");
        assert_cli!("install", "shfmt@2");
        let stdout = assert_cli!("local", "shfmt@2");
        assert_snapshot!(stdout);
        let stdout = assert_cli!("local", "--fuzzy", "shfmt@2");
        assert_snapshot!(stdout);
        let stdout = assert_cli!("local", "--remove", "nodejs");
        assert_snapshot!(stdout);
        let stdout = assert_cli!("ls", "--current");
        assert_str_eq!(
            grep(stdout, "nodejs"),
            "   nodejs     18.0.0 (missing)   (set by ~/cwd/.node-version)"
        );

        fs::write(cf_path, orig).unwrap();
    }
}
