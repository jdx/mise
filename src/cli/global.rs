use std::path::PathBuf;

use eyre::Result;

use crate::cli::args::tool::{ToolArg, ToolArgParser};
use crate::cli::local::local;
use crate::config::Config;
use crate::plugins::PluginName;
use crate::{dirs, env};

/// Sets/gets the global tool version(s)
///
/// Displays the contents of ~/.tool-versions after writing.
/// The file is `$HOME/.tool-versions` by default. It can be changed with `$RTX_CONFIG_FILE`.
/// If `$RTX_CONFIG_FILE` is set to anything that ends in `.toml`, it will be parsed as `.rtx.toml`.
/// Otherwise, it will be parsed as a `.tool-versions` file.
/// A future v2 release of rtx will default to using `~/.config/rtx/config.toml` instead.
///
/// Use `rtx local` to set a tool version locally in the current directory.
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, hide = true, alias = "g", after_long_help = AFTER_LONG_HELP)]
pub struct Global {
    /// Tool(s) to add to .tool-versions
    /// e.g.: node@20
    /// If this is a single tool with no version, the current value of the global
    /// .tool-versions will be displayed
    #[clap(value_name = "TOOL@VERSION", value_parser = ToolArgParser, verbatim_doc_comment)]
    tool: Vec<ToolArg>,

    /// Save exact version to `~/.tool-versions`
    /// e.g.: `rtx global --pin node@20` will save `node 20.0.0` to ~/.tool-versions
    #[clap(long, verbatim_doc_comment, overrides_with = "fuzzy")]
    pin: bool,

    /// Save fuzzy version to `~/.tool-versions`
    /// e.g.: `rtx global --fuzzy node@20` will save `node 20` to ~/.tool-versions
    /// this is the default behavior unless RTX_ASDF_COMPAT=1
    #[clap(long, verbatim_doc_comment, overrides_with = "pin")]
    fuzzy: bool,

    /// Remove the plugin(s) from ~/.tool-versions
    #[clap(long, value_name = "PLUGIN", aliases = ["rm", "unset"])]
    remove: Option<Vec<PluginName>>,

    /// Get the path of the global config file
    #[clap(long)]
    path: bool,
}

impl Global {
    pub fn run(self) -> Result<()> {
        let config = Config::try_get()?;
        local(
            &config,
            &global_file(),
            self.tool,
            self.remove,
            self.pin,
            self.fuzzy,
            self.path,
        )
    }
}

fn global_file() -> PathBuf {
    env::RTX_CONFIG_FILE.clone().unwrap_or_else(|| {
        if *env::RTX_USE_TOML {
            dirs::CONFIG.join("config.toml")
        } else {
            dirs::HOME.join(env::RTX_DEFAULT_TOOL_VERSIONS_FILENAME.as_str())
        }
    })
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
  # set the current version of node to 20.x
  # will use a fuzzy version (e.g.: 20) in .tool-versions file
  $ <bold>rtx global --fuzzy node@20</bold>

  # set the current version of node to 20.x
  # will use a precise version (e.g.: 20.0.0) in .tool-versions file
  $ <bold>rtx global --pin node@20</bold>

  # show the current version of node in ~/.tool-versions
  $ <bold>rtx global node</bold>
  20.0.0
"#
);

#[cfg(test)]
mod tests {
    use crate::{dirs, file};
    use pretty_assertions::assert_str_eq;

    #[test]
    fn test_global() {
        let cf_path = dirs::HOME.join(".test-tool-versions");
        let orig = file::read_to_string(&cf_path).ok();
        let _ = file::remove_file(&cf_path);

        assert_cli!("install", "tiny@2");
        assert_cli_snapshot!("global", "--pin", "tiny@2");
        assert_cli_snapshot!("global", "tiny@2");
        assert_cli_snapshot!("global", "--remove", "tiny");
        assert_cli_snapshot!("global", "--pin", "tiny", "2");

        // will output the current version(s)
        assert_cli_snapshot!("global", "tiny");

        // this plugin isn't installed
        let err = assert_cli_err!("global", "invalid-plugin");
        assert_str_eq!(
            err.to_string(),
            "no version set for invalid-plugin in ~/.test-tool-versions"
        );

        // can only request a version one plugin at a time
        let err = assert_cli_err!("global", "tiny", "dummy");
        assert_str_eq!(err.to_string(), "invalid input, specify a version for each tool. Or just specify one tool to print the current version");

        // this is just invalid
        let err = assert_cli_err!("global", "tiny", "dummy@latest");
        assert_str_eq!(err.to_string(), "invalid input, specify a version for each tool. Or just specify one tool to print the current version");

        assert_cli_snapshot!("global", "--path");

        if let Some(orig) = orig {
            file::write(cf_path, orig).unwrap();
        }
    }
}
