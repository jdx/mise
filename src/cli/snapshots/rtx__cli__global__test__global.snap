---
source: src/cli/global.rs
expression: stdout
---
shfmt 2.6.4
